keyvi/src/cpp/compression/compression_strategy.h

```diff
@@ -0,0 +1,95 @@
+/* * keyvi - A key value store.
+ *
+ * Copyright 2015 Hendrik Muhs<hendrik.muhs@gmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * compression_strategy.h
+ *
+ *  Created on: October 1, 2015
+ *      Author: David Mark Nemeskey<nemeskey.david@gmail.com>
+ */
+
+#ifndef COMPRESSION_STRATEGY_H_
+#define COMPRESSION_STRATEGY_H_
+
+#include <string>
+
+namespace keyvi {
+namespace compression {
+
+enum CompressionCode {
+  NO_COMPRESSION = 0,
+  ZLIB_COMPRESSION = 1,
+  SNAPPY_COMPRESSION = 2,
+};
+
+/**
+ * The base class of every compression strategy.
+ *
+ * All strategies (aside from RawCompressionStrategy) should insert a " "
+ * (space) character before the string to mark it compressed, as well as the
+ * encoded length of the resulting string (including the space). The space,
+ * but not the encoded size, will also have to be taken into account in the
+ * decompress method.
+ */
+struct CompressionStrategy {
+  virtual ~CompressionStrategy() = default;
+
+  inline std::string Compress(const std::string& raw) {
```

Comment:
what if the output is not a string but the output is written into a std::ostream. That would allow buffer-reusage. If you want a string you can pass in a stringstream. We can also define convenience methods here: CompressToString and DecompressToString().